File: datastructures/src/main/java/com/javaground/datastructures/linkedlists/singly/SinglyLinkedListNode.java
package com.javaground.datastructures.linkedlists.singly;

import com.javaground.datastructures.linkedlists.LinkedListNode;

/**
 * Represents a node in a singly linked list structure. Each node contains
 * data of a generic type and a reference to the next node in the list.
 *
 * @param <T> the generic type of the data stored in the node, which must
 *            implement the Comparable interface.
 */
public class SinglyLinkedListNode<T extends Comparable<T>> extends LinkedListNode<T> {
    public SinglyLinkedListNode<T> next = null;

    public SinglyLinkedListNode(T data) {
        super(data);
    }

    public SinglyLinkedListNode(T data, SinglyLinkedListNode<T> next) {
        super(data, next);
    }

    public SinglyLinkedListNode(T data, SinglyLinkedListNode<T> next, String key) {
        super(data, next, key);
    }
}

File: datastructures/src/main/kotlin/com/kotlinground/datastructures/linkedlists/singly/SinglyLinkedListNode.kt
package com.kotlinground.datastructures.linkedlists.singly

import com.kotlinground.datastructures.linkedlists.LinkedListNode
import java.security.MessageDigest

/**
 * Representation of a SinglyLinkedList Node
 */
data class SinglyLinkedListNode<T>(
    override var data: T,
    override var next: SinglyLinkedListNode<T>? = null,
    override val key: String? = lazy {
        val md = MessageDigest.getInstance("MD5")
        val digest = md.digest(data.toString().toByteArray())
        digest.toString()
    }.toString()
) : LinkedListNode<T>
